Use a class for help command and do some fixes to tests

voxsim opened the pull request:
This is my first tiny step towards a Base Command class to clash the number of ifs in src/cli/index.js. The way is quite long :D

**Summary**
For now I fixed some tests and write the help command as es6 class.

**Test plan**
I fixed some tests before touching some code; please see the commits.
Added tests for cache command.

@bestander I hope you like what I do and I have some questions for you:
1. Do you think is useful add an acceptance/feature test per command? It maybe be overkill but the code is too much entagled and if you agree we can remove in a second moment after the story ends if they are useless.
2. I think I will open an issue and maybe I will do a checklist because it is quite long and maybe it is useful some feedbacks/help, what do you think?

Right I am working to have all classes with three static functions (run, hasWrapper, setFlags), and when I will be ready everything will be switched to instance function. I will open more pull request to be careful ;)

bestander commented on src/util/misc.js:
```diff
@@ -62,3 +62,13 @@ export function camelCase(str: string): ?string {
   }
 }
+
+export function findIndex(args: Array<string>, predicate: string): number {
```
Why don't you use `args.findIndex()`?

voxsim replied:
Ops you are right ^^' I really don't remember why did this

bestander commented:
@voxsim, thanks for bringing this up.
Let's open an issue to track gradual improvement.
This may require a few PRs, we need to split it into atomic steps.

bestander commented:
Ping me when it is ready for review

voxsim commented:
@bestander I think the code is ready for review; the failure of travis does not seem related to my last changes O.o

voxsim commented:
I added other commits because I had time and maybe you can give me some feedbacks on how I arrange the cache command class (before it was using buildSubCommand, maybe it is worse than before for you).

bestander commented on __tests__/index.js:
```diff
@@ -204,3 +204,9 @@ if (process.platform !== 'win32') {
+
+test.concurrent('should run bin command', async () => {
+  const stdout = await execCommand('bin', [], '', false);
+  expect(stdout[0]).toEqual(`${fixturesLoc}/node_modules/.bin`);
```
this test fails on windows CI (appveyor)

voxsim replied:
Oh what a newbie error ^^' i'll fix asap!

voxsim commented:
@bestander AppVeyor breaks for both builds for:

`Timeout - Async callback was not invoked within timeout specified by jasmine.DEFAULT_TIMEOUT_INTERVAL`

That does not seem related to my changes (I haven't changed anything related to the installation command). Could you please check? I'll investigate further in the next days.

I added some tests for the cache command, now every change is tested properly!

voxsim commented:
Maybe this pull request https://github.com/yarnpkg/yarn/pull/2887 is better. I will close this after discussing which is better.

bestander commented:
I'll assign @arcanis as reviewer of this and the alternative PR.
Good work, @voxsim, sorry for the delays, we are just trying to balance between fixing bugs, adding new features and improving code maintainability.
Sometimes the latter gets a lower priority.

bestander commented:
Overall I like the BaseCommand/SubCommand idea but I let @arcanis drive this with you

voxsim commented:
after some parallel work i want to address this in another way ;) tomorrow or sunday I will write something here or write another pull request.

Btw do you need some help about addressing issues or bugs? I can help, sorry for the last month but I was busy for personal reasons, but I should come back :p

bestander commented:
Yes yes yes! :)

We definitely need help with triagibg bugs and reviewing PRs.
Just go ahead and review PRs and ping me for merges or access to labelling.

We usually give permissions to active members after they've sent a dozen
PRs and shown care for the project.


On Fri, 7 Apr 2017 at 19:42, Simon Vocella <notifications@github.com> wrote:

> after some parallel work i want to address this in another way ;) tomorrow
> or sunday I will write something here or write another pull request.
>
> Btw do you need some help about addressing issues or bugs? I can help,
> sorry for the last month but I was busy for personal reasons, but I should
> come back :p
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/yarnpkg/yarn/pull/2810#issuecomment-292619333>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACBdWOUa2lgkFEqo3POTDhzycgthzSQ7ks5rtoOEgaJpZM4MO7NR>
> .
>

voxsim commented:
ok ;)

voxsim commented:
Right now I am closing this pull request in favor of https://github.com/yarnpkg/yarn/pull/3105. Maybe in the future we will extract a base class, we will see :)
